src/protocol_splitter.cpp

```diff
@@ -137,14 +137,24 @@ int DevSerial::open_uart()
 		char aux[64];
 		bool flush = false;
 
-		while (0 < ::read(_uart_fd, (void *)&aux, 64)) {
```

Comment:
instead of trying to read it empty, would a simple tcflush(_uart_fd,TCIOFLUSH); work?

Reply:
The problem to use tcflush in this particular code is that it does not tell whether we flushed some data or was it empty. This code relies to the information when the buffer was not filled anymore during the short 1ms sleep period.
My wild guess is that the flushing code here is meant to work so that it reads the data until there is a short break in data stream indicating that the previous message has just been fully received and we are waiting for next to come. In this point flushing is ended and it starts to receive the data. And "hopefully" next buffer read starts from the very beginning of the next message instead of maybe starting from the middle of it.

Of course this is not reliable way to sync into beginning of message, because the time between messages depends on the baudrate and also the message sending freq in PX4 side. I added that timeout to avoid infinite loop.
Another way would be to just skip the whole flushing (or flush only once) and then just let the protocol header parser to make the job. And in case the payload happen to have correct protocol_splitter header, then the RTPS and/or mavlink parser will discard the message if detects it as garbage.

Reply:
Modified the flushing part to flush only once with tcflush. Leave garbage detecting/discarding for protocol header parser